src/DssGateSuck.sol

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// Copyright (C) 2022 Vamsi Alluri
+// Copyright (C) 2022 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity 0.8.11;
+
+interface VatLike {
+    function live() external returns (uint256);
+    function suck(address, address, uint256) external;
+    function dai(address) external view returns (uint256);
+    function move(address, address, uint256) external;
+}
+
+interface JoinLike {
+    function join(address, uint256) external;
+}
+
+/**
+ @title Gate Suck
+ FEATURES
+  * token approval style draw limit on vat.suck
+
+ DEPLOYMENT
+  * ideally, each gate contract should only be linked to a single integration
+  * authorized integration can then request a dai amount from gate contract with a "suck" call
+
+ DRAW LIMIT
+  * a limit on the amount of dai that can an integration can draw with a vat.suck call
+  * simple gate uses an approved total amount, similar to a token approval
+  * integrations can access up to this dai amount in total
+  * if they repay the dai they can "refill the approval amount" allowing for future draws.
+
+ DAI FORMAT
+  * integrates with dai balance on vat, which uses the dsmath rad number type- 45 decimal fixed-point number
+
+ MISCELLANEOUS
+  * does not execute vow.heal to ensure the dai draw amount from vat.suck is lower than the surplus buffer currently held in vow
+  * vat, and vow addresses cannot be updated after deployment
+*/
+contract DssGateSuck {
+    // --- Auth ---
+    mapping (address => uint256) public wards;                                       // Addresses with admin authority
+    event Rely(address indexed usr);
```

Comment:
Suggest to have all the events grouped together in a separated section. It's also good to be able to see the storage layout.

Reply:
Yeah I agree with this, I've been trying to implement that way in new contracts. Group all storage variables, including `wards`, the immutables (with constants), modifiers, events, etc. I feel is much clearer than the original `dss` way.